updates_and_signals/atomic_message_handlers/workflow.py

```diff
@@ -0,0 +1,218 @@
+            # exception from there, or raise an ApplicationError. Other exceptions in the main handler
+            # will cause the workflow to keep retrying and get it stuck.
+            raise ApplicationError("Cannot delete a job: Cluster is already shut down")
+
+        async with self.nodes_lock:
+            nodes_to_free = [
+                k for k, v in self.state.nodes.items() if v == input.task_name
+            ]
+            # This await would be dangerous without nodes_lock because it yields control and allows interleaving.
+            await self._deallocate_nodes_for_job(nodes_to_free, input.task_name)
+
+    async def _deallocate_nodes_for_job(self, nodes_to_free: List[str], task_name: str):
+        await workflow.execute_activity(
+            deallocate_nodes_for_job,
+            DeallocateNodesForJobInput(nodes=nodes_to_free, task_name=task_name),
+            start_to_close_timeout=timedelta(seconds=10),
+        )
+        for node in nodes_to_free:
+            self.state.nodes[node] = None
+
+    def get_assigned_nodes(self) -> List[str]:
+        return [k for k, v in self.state.nodes.items() if v is not None and v != "BAD!"]
+
+    async def perform_health_checks(self):
+        async with self.nodes_lock:
+            assigned_nodes = self.get_assigned_nodes()
+            # This await would be dangerous without nodes_lock because it yields control and allows interleaving.
+            bad_nodes = await workflow.execute_activity(
+                find_bad_nodes,
+                FindBadNodesInput(nodes_to_check=assigned_nodes),
+                start_to_close_timeout=timedelta(seconds=10),
+                # This health check is optional, and our lock would block the whole workflow if we let it retry forever.
+                retry_policy=RetryPolicy(maximum_attempts=1),
+            )
+            for node in bad_nodes:
+                self.state.nodes[node] = "BAD!"
+
+    # The cluster manager is a long-running "entity" workflow so we need to periodically checkpoint its state and
+    # continue-as-new.
+    def init(self, input: ClusterManagerInput):
+        if input.state:
+            self.state = input.state
+        if input.test_continue_as_new:
+            self.max_history_length = 120
+            self.sleep_interval_seconds = 1
+        else:
+            self.max_history_length = None
+            self.sleep_interval_seconds = 600
+
+    def should_continue_as_new(self):
+        # We don't want to continue-as-new if we're in the middle of an update
+        if self.nodes_lock.locked():
+            return False
+        if workflow.info().is_continue_as_new_suggested():
+            return True
+        # This is just for ease-of-testing.  In production, we trust temporal to tell us when to continue as new.
+        if (
+            self.max_history_length
+            and workflow.info().get_current_history_length() > self.max_history_length
+        ):
+            return True
```

Comment:
This makes it more confusing from a pedagogical point of view. Might be nice to switch to e.g. using `mock.patch` in the test to control CAN limit. (Non-blocking comment)

Reply:
Note that it's not just a pytest affordance, it's also for the sample.  (there's a --test-continue-as-new argument)